crates/next-error-code-swc-plugin/src/lib.rs

```diff
@@ -0,0 +1,275 @@
+    // 1. Git commit hash, concatenated with
+    // 2. the hash of object { file_path, error_message, occurrence_count }
+    fn get_error_code(&mut self, first_arg: &ExprOrSpread) -> String {
+        let error_message: String = self.stringify_new_error_arg(&first_arg.expr);
+
+        // Maintain string_occurrences
+        *self
+            .string_occurrences
+            .entry(error_message.clone())
+            .or_insert(0) += 1;
+
+        // Generate the hash of object { file_path, error_message, occurrence_count }
+        let mut hasher = DefaultHasher::new();
+        serde_json::json!({
+            "file_path": self.file_path,
+            "error_message": error_message,
+            "occurrence_count": self.string_occurrences[&error_message]
+        })
+        .to_string()
+        .hash(&mut hasher);
+        let hash = format!("{:08x}", hasher.finish());
+
+        // Concatenate the commit hash and the hash
+        let error_code = format!("E{}{}", self.commit_hash, hash);
+
+        // Save to JSON file
+        let error_metadata = serde_json::json!({
+            "file_path": self.file_path,
+            "error_message": error_message,
+            "occurrence_count": self.string_occurrences[&error_message]
+        })
+        .to_string();
+        let mut retries = 3;
+
+        // Do not do FS operations in unit tests
+        if cfg!(test) {
+            return error_code;
+        }
+
+        loop {
+            if self.mode == "check" {
+                let error_code_path = format!("cwd/error_codes/{}.json", hash);
+                if !std::path::Path::new(&error_code_path).exists() {
+                    panic!(
+                        "ERROR: File {} does not exist.\n\nREQUIRED ACTION:\n1. Run `pnpm \
+                         build`\n2. Commit all file changes from \
+                         /packages/next/error-codes\n\nThis is required to maintain error code \
+                         consistency.",
+                        format!("/packages/next/error_codes/{}.json", hash)
+                    );
+                }
+                return error_code;
+            }
+
+            if self.mode != "generate" {
+                panic!("Mode must be 'generate', got '{}'", self.mode);
+            }
+
+            fs::create_dir_all("cwd/error_codes").unwrap_or_else(|e| {
+                panic!("Failed to create error_codes directory: {}", e);
+            });
+
+            match fs::write(format!("cwd/error_codes/{}.json", hash), &error_metadata) {
+                Ok(_) => break,
+                Err(e) => {
+                    retries -= 1;
+                    if retries == 0 {
+                        panic!("Failed to write error metadata after 3 retries: {}", e);
+                    }
+                }
+            }
+        }
+
+        error_code
+    }
+}
+
+impl VisitMut for TransformVisitor {
+    fn visit_mut_expr(&mut self, expr: &mut Expr) {
```

Comment:
it'd probably be more efficient to use [`visit_mut_new_expr`](https://rustdoc.swc.rs/swc_ecma_visit/trait.VisitMut.html#method.visit_mut_new_expr), that way SWC handles the filtering for us

Reply:
The visit_mut_new_expr approach didn't work because we can't transform a NewExpr into a CallExpr directly - we need to operate at the `Expr` level where we can swap between variants.

Reply:
ah TIL! that's unfortunate